### ISSUE TITLE
KTLS Support

### ISSUE THREAD

**Issue by quininer** [rustls/rustls#198]
KTLS is tls support in linux kernel, but it only handles symmetric encryption and decryption, so using Rustls as a handshake is great.

We can simply public `SessionSecrets`, but this does not apply to TLS 1.3.

**Comment by raphaelcohn** [comment]
This is a big win for servers that regularly serve large files, as it enables to use `sendfile()` and friends once again - and so transfer data from files to the client with zero copy - ie not having to pass copies of bytes through the kernel (and so also significantly reducing the number of syscalls required - another win post-Spectre).

**Comment by thalesfragoso** [comment]
I would like to bring attention to this issue once more,  It seems that there was some hold back because it could make the library "less secure", but how about if we only expose it behind a feature flag like `dangerous_configuration` ?

**Comment by djc** [comment]
Ideally we want some interface here that doesn't require leaking the secrets out of rustls. I have not worked with KTLS before, but would be open to reviewing a proposal on how to make this work.

**Comment by Geal** [comment]
maybe this could be done by integrating https://github.com/quininer/ktls inside rustls? Then having `ServerSession` convert to it?

**Comment by Geal** [comment]
so, after looking a bit more into it, I have a rough plan.
For some context: the way KTLS works is that the userland TLS implementation performs the handshake, then a KTLS file descriptor is created, then the IV, keys and sequence ids are extracted from the TLS session, and added to the KTLS fd (see https://github.com/quininer/ktls/blob/master/src/sys/mod.rs#L40-L62 ). Then read(), write() can be called directly on this fd.

We want to avoid making rustls session secrets easily accessible, so the idea would be to transform a `ClientSession` or `ServerSession` and the associated socket into a `KtlsStream`, which would be provided by rustls too, so the secrets can be transmitted and set up directly, with no easy access.

Possible issues:
- Linux only (I have not looked yet into FreeBSD's ktls implementation)
- limited cipher suites
- in async reactors like tokio or async-std, that rely on epoll, I think the new socket does not need to be registered to the epoll implementation, but I am not sure yet. If reading from the ktls fd consumes everything from the TCP socket's buffer, the kernel should understand that we want to read again? How do we tell the executor which task should be waked? The solution might be to register the ktls fd to epoll. This will require some tests
- who owns which file descriptors, should we forget about the socket's fd after transforming it to a ktls fd?

By reusing the previous ktls proof of concept, I should be able to have a test up and running quickly, if that scheme looks fine. WDYT @djc ?

**Comment by quininer** [comment]
I prefer rustls to provide an api to access secrets, so that rustls keep platform-independent and no unsafe code.

We already have `KeyLog` that will leak secrets, and I believe that exposing another api will not pose additional risks.

**Comment by djc** [comment]
@Geal this mostly sounds good to me! So maybe we could have a method `fn kernel_tls(&self) -> Result<std::net::TcpStream, Error>` (guarded by appropriate platform `cfg` attributes) which validates that the negotiated cipher suite matches the cipher suites supported by the kernel. By yielding a `TcpStream`, we leave any async issues to the tokio or async-std implementations (assuming async-std has something similar to tokio's `TcpStream::from_std()`). Remember that Rustls connections today are external to whatever carries the file descriptors. Maybe `kernel_tls()` should then take an `AsRawFd` as an argument if we need to make sure the old one doesn't survive?

@quininer I don't see a big issue with adding some platform-independent functionality to Rustls as long as it will generate compile-time errors on platforms where it isn't available. I hadn't thought about unsafe code yet, that's a more interesting problem. Maybe the ktls implementation should live in a different crate after all, and provide its own implementation of the `KeyLog` trait to gain access to the required secrets?

**Comment by quininer** [comment]
@djc  If you want to split the ktls code into another crate, it means that rustls must expose secrets in some form. I don't think they are much different, but we can indeed design an interface that is more difficult to abuse.

**Comment by djc** [comment]
As I suggested I think we could reuse the existing `KeyLog` interface for that at least for now.

**Comment by Geal** [comment]
So I guess the two plans are:
- integrate ktls in rustls and create a TcpStream from a session and a file descriptor
- expand KeyLog to provide the sequence ids and IV, and implement ktls in another crate

What was the original purpose of KeyLog? Debug, or decryption elsewhere in the infra?

**Comment by quininer** [comment]
@Geal  I believe it is to implement [Key Log](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format).

**Comment by djc** [comment]
Ah, if we have to expand `KeyLog` it might not be so attractive to use it.

**Comment by Geal** [comment]
So, shall I attempt an implementation of the other idea then?

**Comment by djc** [comment]
Sounds good, then we can figure out how we integrate this exactly.

**Comment by raphaelcohn** [comment]
@djc Whatever we do here, I'd like the ability to use raw file descriptors. I don't use tokio - it's simply not appropriate to my high performance use cases. I do use a custom wrapper around io-uring\* for which having kernel native TLS support might be a win - assuming I can manage the complexity.

\* Which I prefer to tokio for a coroutine programming model.

**Comment by Geal** [comment]
ktls with io-uring sounds awesome

**Comment by raphaelcohn** [comment]
@Geal it may not be possible, but it might be. My custom io-uring wrapper is in the unpublished [linux-support](https://github.com/lemonrock/linux-support) crate. Once I've proved in out with a public DNS implementation, it'll be good to go.

**Comment by djc** [comment]
> @djc Whatever we do here, I'd like the ability to use raw file descriptors. I don't use tokio - it's simply not appropriate to my high performance use cases. I do use a custom wrapper around io-uring* for which having kernel native TLS support might be a win - assuming I can manage the complexity.

rustls (transitively) never holds any file descriptors, so I don't see how that comment is relevant to rustls.

**Comment by raphaelcohn** [comment]
@djc There was talk above about wrapping a file descriptor, and about integrations with tokio - I was simply reacting to that. I'm glad that tokio still isn't integrated into this crate!

**Comment by AlexZhao** [comment]
@briansmith  Any idea will there be some pub interface from rustls can be used or planned to have this type of functionalities? currently seems Linux has kTLS, but I think quite longtime ago there is already discussion about this in BSD like systems. 
it maybe quite useful for high performance usage, both can save moving data across kernel/user space, and direct improve traffic performance.

**Comment by fasterthanlime** [comment]
I built a working demo (2022 edition) of rustls+kTLS — it supports TLS 1.2, TLS 1.3, and all the ciphers I could find:

  * https://github.com/hapsoc/ktls

(Thanks to @quininer for the initial research ✨)

Here's the crimes I committed against rustls to get everything working: https://github.com/rustls/rustls/compare/main...hapsoc:rustls:export

I'm not opening a PR with these changes - they're not up to the standard of the project, etc. etc. However, I would really _really_ like there to be an equivalent interface (giving seq numbers / key / iv / salt, for both sides) upstream. Personally I don't believe making things `pub(crate)` enhances security in some way, but maybe we can find a compromise?

Here's some ideas:

  * gate it with a cargo feature flag, call it `i_realize_exporting_secrets_is_dangerous`
  * instead of returning a struct with some `Vec<u8>` (allocations! shock! awe!), take a callback function that takes a borrowed struct (with u8 slices). this puts the emphasis on the data being radioactive, and that it should be copied to some other secure location as soon as possible, and then forgotten.
  * a third, even better option

I'm keen to hear everyone's thoughts on this — I like my forks short-lived.

**Comment by carbotaniuman** [comment]
I do believe that there should be a public interface for getting the secrets out of RustTLS, for a server I've been working I simply pin RustTLS and transmute all the secrets I need out via wrapper structs, which is horribly unsafe but the only option I can use.

**Comment by djc** [comment]
@carbotaniuman have you explained your use case in the issue tracker before?

**Comment by carbotaniuman** [comment]
It's effectively #424, but my impression was that was dismissed offhandly, so I just worked around it.

**Comment by djc** [comment]
@fasterthanlime do you think it's feasible to have an API where a `RawFd` is passed in, instead of getting the secrets out? If not, why not?

**Comment by fasterthanlime** [comment]
> @fasterthanlime do you think it's feasible to have an API where a `RawFd` is passed in, instead of getting the secrets out? If not, why not?

It is! I love that idea. And not _just_ for the fact that when the BSD people come a-knocking, they'll be your problem, not mine! (Just kidding, love you BSD friends)

In all seriousness - yes, it's possible.

Here are design constraints to keep in mind:

## Falling back to userland encryption

If the `tls` kernel module isn't loaded/enabled, [setting TLS as the ULP protocol](https://github.com/hapsoc/ktls/blob/7d2a6b233258f3114f8f321ca627e17303399f78/src/ffi.rs#L30-L45) will fail (with "file not found", because errno, lol).

If it fails, it's okay! You have rustls already! rustls can do all the things you need to do. You just can't use kTLS. The fd is still good, it's still very probably a TCP stream to some peer, you just can't have the kernel's help.

## Leaving the fd in a bad state

If [setting TLS info](https://github.com/hapsoc/ktls/blob/7d2a6b233258f3114f8f321ca627e17303399f78/src/ffi.rs#L221-L227) (passing secrets to the kernel) fails, I think the fd is in a bad state. It's set as using TLS as the ULP protocol, but there's no secrets. Then again, you need to set up the TX / RX directions independently already, so possibly it's fine? It just... goes through unnecessary layers?

It might be possible to "uninstall" TLS as the ULP, which might be desirable if this part fails.

However, ideally, we'd want to fail earlier, that's why my lil' ktls crate first checks that the ciphers are supported and everything is good to go _before_ calling the syscalls, by attempting a couple [fallible conversions](https://github.com/hapsoc/ktls/blob/7d2a6b233258f3114f8f321ca627e17303399f78/src/lib.rs#L54-L57), and if fails at that stage, no syscalls are made at all.

In short: we want to prepare everything we need, and _only then_ issue those 3 syscalls in quick succession.

## What to return?

Since the proposed rustls API would operate on a `RawFd`, it should probably return `Ok(())` in case of success, and `Err(CanIStillUseThisFd)` in case of failure.

## Control messages

I had many questions about kTLS before starting this: how does the handshake happen? (turns out you do that in userland and pass secrets to the kernel). One question I should've had, but didn't, was: how do you send/receive control messages? And the answer is ancillary crap via `sendmsg` / `recvmsg` (also iovecs because reasons?).

The good news is that this is probably something for `tokio-rustls` or equivalent to worry about. I have all the right things in my ktls crate, ie. a [KtlsStream](https://github.com/hapsoc/ktls/blob/7d2a6b233258f3114f8f321ca627e17303399f78/src/ktls_stream.rs) type that sends `close_notify` on shutdown/drop - but again, this is out of scope for rustls.

## Which side are you on?

The kTLS APIs think in terms of "tx" (transmit) and "rx" (receive) sides, whereas rustls (and I think, TLS 1.2/1.3) think in terms of client / server - this can be confusing. I seem to have stumbled upon the right combination in the ktls crate, so it seems like an okay reference.

## How to test this

I don't know how rustls's CI is set up currently, but the bad news is you can't test kTLS stuff in a docker container. I don't know if the default GitHub Actions runners have the `tls` module available, but my plan up until now was set up self-hosted runners that are microVMs (that do have the `tls` kernel module).

I suppose the control message stuff (sending close_notify) can be left out of the rustls test suite entirely, if we rely on fixed-length payloads instead (instead of doing `read_to_end` like I do in ktls's tests).

--

Going ahead with a `configure_this_fd_plz`-like API in rustls seems feasible and it would allow the secrets to never escape the rustls crate. However, I'm still not sure as to the reasons for the reluctance to exposing secrets via an API - surely between this and transmuting to equivalent struct definitions just to implement another protocol, this is the lesser evil.

(I understand that good crate design should help folks design safer systems altogether, but that's why there's already a `dangerous_configuration` Cargo feature for example — sometimes, that really is what you want).

If the "configure an FD" API lands in rustls, I'm a happy camper, but it seems like a larger maintenance burden for the rustls project than an "I accept the consequences of my actions" escape hatch — I'm happy to assume the burden of maintaining the separate ktls crate, which can be as platform-specific as it wants, and have an annoying CI setup the rest of rustls doesn't need.

**Comment by djc** [comment]
Yeah, okay, I think I'm convinced that we actually want a way to expose the secrets. So I think we'll want a two-pronged approach: both a Cargo feature (let's asume `extract-secrets` for now -- not sure we want to reuse `dangerous_configuration` for this) and by also setting a value in the `ClientConfig`/`ServerConfig` -- let's call it `enabled_secret_extraction`, and then have an API that allows extracting secrets without incurring any unnecessary allocations as `ConnectionCommon::extract_secrets() -> Option<>`.

The option to do it with a callback setup that minimizes the way secrets can leak sounds appealing, but in practice I'm not sure the callback option provides many extra guarantees compared to a more conventional setup that is kept lifetime-bound.

**Comment by fasterthanlime** [comment]
@djc is there somewhere we can discuss this synchronously? I'd like to start working on such a PR today.

I was thinking about how to extract secrets without unnecessary allocations, but this seems hard because:

 * Keys can have different lengths
 * Salt/IV look different (but so far, together, they all fit within 12 bytes max)

This is all solvable by essentially returning an `enum` with one variant per `BulkAlgorithm` (it would have the size of the largest variant, if we want to avoid any boxing). I wonder how much of my current extract code is ktls-specific, and if it it would be suitable for other usecases 🤔

**Comment by djc** [comment]
I'm hanging out in the rustls channel on Discord, see the README.

**Comment by djc** [comment]
@carbotaniuman so was your use case also for kTLS? See the latest discussion in #1110.

**Comment by carbotaniuman** [comment]
Sorry for the late reply, but no and yes. The context is morally the same, but it's not kTLS but interfacing with some propietary protocol library.

 #1112 is pretty much exactly what I need though, no complaints there!

**Comment by sprappcom** [comment]
can anyone show an example on how to use ktls with rustls?
